curated_transformers/util/serde.py

```diff
@@ -0,0 +1,311 @@
+    # dicts as the process could potentially change the structure of sub-modules,
+    # e.g: when quantized layers rename their parameters.
+    module_keys = set(model.state_dict().keys())
+    seen_keys: Set[str] = set()
+
+    # Ideally, we'd lazily load and convert the state dicts in the incoming
+    # order, but some of the conversion operations induce dependencies b'ween
+    # parameters in different state dicts. So, we need to sort them into buckets
+    # before performing the conversion ops, precluding us from "streaming" the
+    # parameters into the model. Consequently, this requires us to hold the
+    # incomplete buckets (and their parameters) in memory until all of their
+    # parameters have been accumulated.
+    default_bucket = DefaultParamBucket()
+    for state_dict in state_dicts:
+        # Sort into buckets.
+        for name, param in state_dict.items():
+            matching_buckets = [
+                bucket for bucket in deserialization_buckets if bucket.match(name)
+            ]
+            num_matching_buckets = len(matching_buckets)
+            if num_matching_buckets == 0:
+                default_bucket[name] = param
+            elif num_matching_buckets == 1:
+                matching_buckets[0][name] = param
+            else:
+                raise ValueError(
+                    f"Key `{name}` matched multiple ({num_matching_buckets}) deserialization buckets"
+                )
+
+        # Convert and load the default bucket first.
+        convert_and_load_state_dict(default_bucket, seen_keys)
+        default_bucket.clear()
+
+        # Process other buckets that are full.
+        for bucket in deserialization_buckets:
+            if bucket.ready():
+                convert_and_load_state_dict(bucket, seen_keys)
+                bucket.clear()
+
+    # Make sure that we didn't miss any keys.
+    missing_keys = module_keys.difference(seen_keys)
+    if len(missing_keys) != 0:
+        raise ValueError(f"Some parameters were not updated/replaced: {missing_keys}")
+
+    if any(len(bucket) for bucket in deserialization_buckets):
+        raise ValueError("One or more deserialization buckets were not empty")
+
+
+def default_tensor_to_parameter_converter(
+    module: Module,
+    module_prefix: str,
+    parameter_name: str,
+    tensor: torch.Tensor,
+    device: Optional[torch.device] = None,
+) -> Parameter:
+    """Default tensor to parameter converter.
+
+    :param module:
+        Parent module of the parameter being converted/replaced.
+    :param module_prefix:
+        Prefix of the parent module.
+    :param parameter_name:
+        Name of the parameter being converted/replaced.
+    :param tensor:
+        Tensor to be converted.
+    :param device:
+        Device to which the converted parameter is moved.
+    :returns:
+        Converted parameter.
+    """
+    old_param = module._parameters[parameter_name]
```

Comment:
Can we use

https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.get_parameter

here rather than relying on internals?

Reply:
We could, but we'd need to reconstruct the fully-qualified key using both `module_prefix` and `parameter_name` to pass to `get_parameter`. Also, we still have to manually modify the `_parameters` dict when we replace the old parameter with the new one, so the reliance on internals will still be there.

Reply:
Ok, let's leave it like this, since it's to have the bucketing code in git.